File: ios-native/KyouNoOgatore/KyouNoOgatore/QuizArt.swift
//
//  QuizArt.swift
//  KyouNoOgatore
//
//  TASK-C2-2026-07-28-quiz-result-reach-parity.md §4: app-quiz.js:92-137 QUIZ_ART[2]/[3]の移植。
//  Web版コメント「あたま/あごの高さ線＝両ひじが上がる目安」「かかとの浮き」は判定基準そのものの
//  可視化であり装飾ではない。SVGの正確な手描き曲線までは再現せず、判定基準(高さの目安線・
//  かかとの浮き)が明確に伝わる簡略化した図として実装する(タスク文が明示的に許容する範囲)。
//  Android版QuizArt.ktと同一ロジック。

import SwiftUI

private let quizArtInk = Color(hex: 0x3A3A35)
private let quizArtSkin = Color(hex: 0xE8B48C)
private let quizArtPink = Color(hex: 0xE56A9A)
private let quizArtHead = Color(hex: 0xFFE3C9)
private let quizArtGround = Color(hex: 0xE0D8C4)
private let quizArtGuide = Color(hex: 0xCFC9B8)
private let quizArtLabel = Color(hex: 0x6E6B5F)

struct QuizArtKenko: View {
    var body: some View {
        Canvas { ctx, size in
            let s = size.width / 300
            func pt(_ x: CGFloat, _ y: CGFloat) -> CGPoint { CGPoint(x: x * s, y: y * s) }

            // app-quiz.js:96 M30 150h240(ゆか)
            var ground = Path()
            ground.move(to: pt(30, 150)); ground.addLine(to: pt(270, 150))
            ctx.stroke(ground, with: .color(quizArtGround), style: StrokeStyle(lineWidth: 5 * s, lineCap: .round))

            // app-quiz.js:97-99 はな/あごの高さの目安線(点線)+ラベル。判定基準そのもの。
            var noseLine = Path(); noseLine.move(to: pt(136, 57)); noseLine.addLine(to: pt(210, 57))
            var chinLine = Path(); chinLine.move(to: pt(136, 69)); chinLine.addLine(to: pt(210, 69))
            let dash = StrokeStyle(lineWidth: 2.5 * s, dash: [2 * s, 6 * s])
            ctx.stroke(noseLine, with: .color(quizArtGuide), style: dash)
            ctx.stroke(chinLine, with: .color(quizArtGuide), style: dash)
            ctx.draw(Text("はな").font(.system(size: 11, weight: .black)).foregroundColor(quizArtLabel), at: pt(226, 58), anchor: .leading)
            ctx.draw(Text("あご").font(.system(size: 11, weight: .black)).foregroundColor(quizArtLabel), at: pt(226, 71), anchor: .leading)

            // app-quiz.js:100 legs
            var legL = Path(); legL.move(to: pt(112, 118)); legL.addLine(to: pt(106, 150))
            var legR = Path(); legR.move(to: pt(132, 118)); legR.addLine(to: pt(138, 150))
            let legStyle = StrokeStyle(lineWidth: 9 * s, lineCap: .round)
            ctx.stroke(legL, with: .color(Color(hex: 0x55524A)), style: legStyle)
            ctx.stroke(legR, with: .color(Color(hex: 0x55524A)), style: legStyle)
            // app-quiz.js:101 torso
            var torso = Path(); torso.move(to: pt(122, 118)); torso.addLine(to: pt(122, 80))
            ctx.stroke(torso, with: .color(quizArtInk), style: StrokeStyle(lineWidth: 15 * s, lineCap: .round))
            // app-quiz.js:102 head
            let head = Path(ellipseIn: CGRect(x: pt(122, 58).x - 13 * s, y: pt(122, 58).y - 13 * s, width: 26 * s, height: 26 * s))
            ctx.fill(head, with: .color(quizArtHead))
            ctx.stroke(head, with: .color(quizArtInk), lineWidth: 4 * s)
            // app-quiz.js:105 eyes
            ctx.fill(Path(ellipseIn: CGRect(x: pt(118, 59).x - 2 * s, y: pt(118, 59).y - 2 * s, width: 4 * s, height: 4 * s)), with: .color(quizArtInk))
            ctx.fill(Path(ellipseIn: CGRect(x: pt(126, 59).x - 2 * s, y: pt(126, 59).y - 2 * s, width: 4 * s, height: 4 * s)), with: .color(quizArtInk))

            // app-quiz.js:107-110 両ひじをつけて上げた腕(肩→ひじ)
            var arms = Path()
            arms.move(to: pt(108, 80))
            arms.addQuadCurve(to: pt(118, 42), control: pt(106, 58))
            arms.move(to: pt(136, 80))
            arms.addQuadCurve(to: pt(126, 42), control: pt(138, 58))
            ctx.stroke(arms, with: .color(quizArtSkin), style: StrokeStyle(lineWidth: 8 * s, lineCap: .round))
            // app-quiz.js:112 ひじが合わさる位置(判定の中心点)
            ctx.fill(Path(ellipseIn: CGRect(x: pt(122, 41).x - 7 * s, y: pt(122, 41).y - 7 * s, width: 14 * s, height: 14 * s)), with: .color(quizArtPink))
        }
        .aspectRatio(300.0 / 170.0, contentMode: .fit)
        .frame(maxWidth: .infinity)
    }
}

struct QuizArtAshi: View {
    var body: some View {
        Canvas { ctx, size in
            let s = size.width / 300
            func pt(_ x: CGFloat, _ y: CGFloat) -> CGPoint { CGPoint(x: x * s, y: y * s) }

            // app-quiz.js:117 ゆか
            var ground = Path(); ground.move(to: pt(30, 150)); ground.addLine(to: pt(270, 150))
            ctx.stroke(ground, with: .color(quizArtGround), style: StrokeStyle(lineWidth: 5 * s, lineCap: .round))

            // app-quiz.js:118-120 しゃがんだ脚(太もも・すね・かかとが浮いた足)
            var thigh = Path(); thigh.move(to: pt(120, 112)); thigh.addLine(to: pt(152, 110))
            var shin = Path(); shin.move(to: pt(120, 112)); shin.addLine(to: pt(134, 142))
            var foot = Path(); foot.move(to: pt(120, 148)); foot.addLine(to: pt(142, 144))
            ctx.stroke(thigh, with: .color(Color(hex: 0x55524A)), style: StrokeStyle(lineWidth: 13 * s, lineCap: .round))
            ctx.stroke(shin, with: .color(Color(hex: 0x55524A)), style: StrokeStyle(lineWidth: 11 * s, lineCap: .round))
            ctx.stroke(foot, with: .color(quizArtInk), style: StrokeStyle(lineWidth: 6 * s, lineCap: .round))
            // app-quiz.js:122 背中
            var back = Path()
            back.move(to: pt(152, 112))
            back.addQuadCurve(to: pt(158, 76), control: pt(150, 88))
            ctx.stroke(back, with: .color(quizArtInk), style: StrokeStyle(lineWidth: 15 * s, lineCap: .round))
            // app-quiz.js:123 head
            let head = Path(ellipseIn: CGRect(x: pt(152, 60).x - 15 * s, y: pt(152, 60).y - 15 * s, width: 30 * s, height: 30 * s))
            ctx.fill(head, with: .color(quizArtHead))
            ctx.stroke(head, with: .color(quizArtInk), lineWidth: 4 * s)
            ctx.fill(Path(ellipseIn: CGRect(x: pt(146, 62).x - 2.2 * s, y: pt(146, 62).y - 2.2 * s, width: 4.4 * s, height: 4.4 * s)), with: .color(quizArtInk))
            // app-quiz.js:126-127 前に伸ばした腕
            var arm = Path(); arm.move(to: pt(156, 80)); arm.addLine(to: pt(118, 86))
            ctx.stroke(arm, with: .color(quizArtSkin), style: StrokeStyle(lineWidth: 8 * s, lineCap: .round))

            // app-quiz.js:128-134 かかとの浮きをズームする円+点線の引き出し線+ラベル(判定基準そのもの)
            var leader = Path(); leader.move(to: pt(158, 140)); leader.addLine(to: pt(190, 144))
            ctx.stroke(leader, with: .color(quizArtGuide), style: StrokeStyle(lineWidth: 2.5 * s, dash: [3 * s, 4 * s]))
            let zoom = Path(ellipseIn: CGRect(x: pt(224, 110).x - 34 * s, y: pt(224, 110).y - 34 * s, width: 68 * s, height: 68 * s))
            ctx.fill(zoom, with: .color(.white))
            ctx.stroke(zoom, with: .color(quizArtInk), lineWidth: 3.5 * s)
            var zoomGround = Path(); zoomGround.move(to: pt(204, 128)); zoomGround.addLine(to: pt(244, 128))
            ctx.stroke(zoomGround, with: .color(quizArtGround), lineWidth: 4 * s)
            var heel = Path(); heel.move(to: pt(209, 124)); heel.addLine(to: pt(235, 118))
            ctx.stroke(heel, with: .color(quizArtInk), style: StrokeStyle(lineWidth: 6 * s, lineCap: .round))
            // かかとが浮いている隙間を示す矢印
            var arrowShaft = Path(); arrowShaft.move(to: pt(234, 120)); arrowShaft.addLine(to: pt(234, 128))
            ctx.stroke(arrowShaft, with: .color(quizArtPink), style: StrokeStyle(lineWidth: 3.5 * s, dash: [2 * s, 3 * s]))
            var arrowHead = Path()
            arrowHead.move(to: pt(229, 122)); arrowHead.addLine(to: pt(234, 128)); arrowHead.addLine(to: pt(239, 122))
            ctx.stroke(arrowHead, with: .color(quizArtPink), style: StrokeStyle(lineWidth: 3 * s, lineCap: .round))
            ctx.draw(
                Text("かかと").font(.system(size: 12, weight: .black)).foregroundColor(quizArtPink),
                at: pt(224, 88)
            )
        }
        .aspectRatio(300.0 / 170.0, contentMode: .fit)
        .frame(maxWidth: .infinity)
    }
}

File: android-native/KyouNoOgatore/app/src/main/java/jp/ogatore/kyouno/QuizArt.kt
package jp.ogatore.kyouno

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.nativeCanvas
import androidx.compose.ui.text.drawText
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.rememberTextMeasurer
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.sp

// TASK-C2-2026-07-28-quiz-result-reach-parity.md §4: app-quiz.js:92-137 QUIZ_ART[2]/[3]の移植。
// Web版コメント「あたま/あごの高さ線＝両ひじが上がる目安」「かかとの浮き」は判定基準そのものの
// 可視化であり装飾ではない(以前は「kenko/ashiは元々手描きSVG/装飾で写真なし」と誤認されて
// 移植対象から外れていた)。SVGの正確な手描き曲線までは再現せず、判定基準(高さの目安線・
// かかとの浮き)が明確に伝わる簡略化した図として実装する(タスク文が明示的に許容する範囲)。
private val quizArtInk = Color(0xFF3A3A35)
private val quizArtSkin = Color(0xFFE8B48C)
private val quizArtPink = Color(0xFFE56A9A)
private val quizArtHead = Color(0xFFFFE3C9)
private val quizArtGround = Color(0xFFE0D8C4)
private val quizArtGuide = Color(0xFFCFC9B8)
private val quizArtLabel = Color(0xFF6E6B5F)

@Composable
fun QuizArtKenko(modifier: Modifier = Modifier) {
    val textMeasurer = rememberTextMeasurer()
    Canvas(modifier.fillMaxWidth().aspectRatio(300f / 170f)) {
        val s = size.width / 300f
        fun pt(x: Float, y: Float) = Offset(x * s, y * s)

        // index.html app-quiz.js:96 M30 150h240(ゆか)
        drawLine(quizArtGround, pt(30f, 150f), pt(270f, 150f), strokeWidth = 5f * s, cap = StrokeCap.Round)

        // app-quiz.js:97-99 はな/あごの高さの目安線(点線)+ラベル。判定基準そのもの。
        val dash = PathEffect.dashPathEffect(floatArrayOf(2f * s, 6f * s))
        drawLine(quizArtGuide, pt(136f, 57f), pt(210f, 57f), strokeWidth = 2.5f * s, pathEffect = dash)
        drawLine(quizArtGuide, pt(136f, 69f), pt(210f, 69f), strokeWidth = 2.5f * s, pathEffect = dash)
        val labelStyle = androidx.compose.ui.text.TextStyle(color = quizArtLabel, fontSize = 11.sp, fontWeight = FontWeight.Black)
        drawText(textMeasurer, "はな", topLeft = pt(215f, 53f), style = labelStyle)
        drawText(textMeasurer, "あご", topLeft = pt(215f, 66f), style = labelStyle)

        // app-quiz.js:100 legs
        drawLine(Color(0xFF55524A), pt(112f, 118f), pt(106f, 150f), strokeWidth = 9f * s, cap = StrokeCap.Round)
        drawLine(Color(0xFF55524A), pt(132f, 118f), pt(138f, 150f), strokeWidth = 9f * s, cap = StrokeCap.Round)
        // app-quiz.js:101 torso
        drawLine(quizArtInk, pt(122f, 118f), pt(122f, 80f), strokeWidth = 15f * s, cap = StrokeCap.Round)
        // app-quiz.js:102 head
        drawCircle(quizArtHead, radius = 13f * s, center = pt(122f, 58f))
        drawCircle(quizArtInk, radius = 13f * s, center = pt(122f, 58f), style = Stroke(4f * s))
        // app-quiz.js:105 eyes
        drawCircle(quizArtInk, radius = 2f * s, center = pt(118f, 59f))
        drawCircle(quizArtInk, radius = 2f * s, center = pt(126f, 59f))

        // app-quiz.js:107-110 両ひじをつけて上げた腕(肩→ひじ)
        val armPath = Path().apply {
            moveTo(pt(108f, 80f).x, pt(108f, 80f).y)
            quadraticBezierTo(pt(106f, 58f).x, pt(106f, 58f).y, pt(118f, 42f).x, pt(118f, 42f).y)
            moveTo(pt(136f, 80f).x, pt(136f, 80f).y)
            quadraticBezierTo(pt(138f, 58f).x, pt(138f, 58f).y, pt(126f, 42f).x, pt(126f, 42f).y)
        }
        drawPath(armPath, quizArtSkin, style = Stroke(8f * s, cap = StrokeCap.Round))
        // app-quiz.js:112 ひじが合わさる位置(判定の中心点)
        drawCircle(quizArtPink, radius = 7f * s, center = pt(122f, 41f))
    }
}

@Composable
fun QuizArtAshi(modifier: Modifier = Modifier) {
    val textMeasurer = rememberTextMeasurer()
    Canvas(modifier.fillMaxWidth().aspectRatio(300f / 170f)) {
        val s = size.width / 300f
        fun pt(x: Float, y: Float) = Offset(x * s, y * s)

        // app-quiz.js:117 ゆか
        drawLine(quizArtGround, pt(30f, 150f), pt(270f, 150f), strokeWidth = 5f * s, cap = StrokeCap.Round)

        // app-quiz.js:118-120 しゃがんだ脚(太もも・すね・かかとが浮いた足)
        drawLine(Color(0xFF55524A), pt(120f, 112f), pt(152f, 110f), strokeWidth = 13f * s, cap = StrokeCap.Round)
        drawLine(Color(0xFF55524A), pt(120f, 112f), pt(134f, 142f), strokeWidth = 11f * s, cap = StrokeCap.Round)
        drawLine(quizArtInk, pt(120f, 148f), pt(142f, 144f), strokeWidth = 6f * s, cap = StrokeCap.Round)
        // app-quiz.js:122 背中
        val backPath = Path().apply {
            moveTo(pt(152f, 112f).x, pt(152f, 112f).y)
            quadraticBezierTo(pt(150f, 88f).x, pt(150f, 88f).y, pt(158f, 76f).x, pt(158f, 76f).y)
        }
        drawPath(backPath, quizArtInk, style = Stroke(15f * s, cap = StrokeCap.Round))
        // app-quiz.js:123 head
        drawCircle(quizArtHead, radius = 15f * s, center = pt(152f, 60f))
        drawCircle(quizArtInk, radius = 15f * s, center = pt(152f, 60f), style = Stroke(4f * s))
        drawCircle(quizArtInk, radius = 2.2f * s, center = pt(146f, 62f))
        // app-quiz.js:126-127 前に伸ばした腕
        drawLine(quizArtSkin, pt(156f, 80f), pt(118f, 86f), strokeWidth = 8f * s, cap = StrokeCap.Round)

        // app-quiz.js:128-134 かかとの浮きをズームする円+点線の引き出し線+ラベル(判定基準そのもの)
        val dash = PathEffect.dashPathEffect(floatArrayOf(3f * s, 4f * s))
        drawLine(quizArtGuide, pt(158f, 140f), pt(190f, 144f), strokeWidth = 2.5f * s, pathEffect = dash)
        drawCircle(Color.White, radius = 34f * s, center = pt(224f, 110f))
        drawCircle(quizArtInk, radius = 34f * s, center = pt(224f, 110f), style = Stroke(3.5f * s))
        drawLine(quizArtGround, pt(204f, 128f), pt(244f, 128f), strokeWidth = 4f * s)
        drawLine(quizArtInk, pt(209f, 124f), pt(235f, 118f), strokeWidth = 6f * s, cap = StrokeCap.Round)
        // かかとが浮いている隙間を示す矢印
        val arrowDash = PathEffect.dashPathEffect(floatArrayOf(2f * s, 3f * s))
        drawLine(quizArtPink, pt(234f, 120f), pt(234f, 128f), strokeWidth = 3.5f * s, pathEffect = arrowDash)
        val arrowHead = Path().apply {
            moveTo(pt(229f, 122f).x, pt(229f, 122f).y)
            lineTo(pt(234f, 128f).x, pt(234f, 128f).y)
            lineTo(pt(239f, 122f).x, pt(239f, 122f).y)
        }
        drawPath(arrowHead, quizArtPink, style = Stroke(3f * s, cap = StrokeCap.Round))
        val labelStyle = androidx.compose.ui.text.TextStyle(color = quizArtPink, fontSize = 12.sp, fontWeight = FontWeight.Black, textAlign = TextAlign.Center)
        val measured = textMeasurer.measure("かかと", labelStyle)
        drawText(textMeasurer, "かかと", topLeft = Offset(pt(224f, 90f).x - measured.size.width / 2f, pt(224f, 90f).y), style = labelStyle)
    }
}
